accepted/automatic-redirects-by-default/automatic-redirects-by-default.md

```diff
@@ -0,0 +1,161 @@
+After some research, she learns that she must at least target .NET Framework
+4.6.1 in order to use .NET Standard 2.0 packages. After upgrading to .NET
+Framework 4.6.1, the package installs without errors. Michelle then starts to
+use pipelines from her app. When she F5s her application it just works fine.
+
+Michelle doesn't know that this is thanks to automatic binding redirect
+generation being on by default. Otherwise her application would have crashed
+with a `FileLoadException`.
+
+### Upgrading a NuGet dependency
+
+Robert is using FizzBuzz, a documentation framework that produces help files out
+of Markdown documents. FizzBuzz depends on Markdig, a popular .NET markdown
+library. Robert encounters an issue with some of his recent documentation
+changes which causes Markdig to crash. He notices that while there is no updated
+version of FizzBuzz, there is an updated version of Markdig that fixes his
+issue. Robert updates his (indirect) dependency on Markdig to the latest version
+that fixes the issue. He triggers a rebuild of the documentation and is happy to
+see that the crash no longer occurs.
+
+Robert doesn't know that this is thanks to automatic binding redirect generation
+being on by default. Otherwise his application would have crashed with a
+`FileLoadException`.
+
+## Requirements
+
+### Goals
+
+* **Turning this setting on must not break working applications**. We expect
+  this setting to make more applications work, but we must not break
+  applications where adding binding redirects breaks them.
+
+* **It must be possible to opt-out**. The customer base on .NET Framework is
+  fairly large and has an unfathomable amount of different configurations. We
+  need to assume that this setting might not work for some extremely rare
+  condition and thus allow those customers to opt-out.
+
+### Non-Goals
+
+* **Introducing new binding policy**. This proposal is only about enabling an
+  already existing feature in MSBuild which uses the also existing binding
+  redirect feature of the CLR.
+
+## Design
+
+Somewhere in the common props we need to add the following snippet:
+
+```xml
+<PropertyGroup>
+    <AutoGenerateBindingRedirects Condition="'$(AutoGenerateBindingRedirects)' == ''">True</AutoGenerateBindingRedirects>
+</PropertyGroup>
+```
+
+Logically, this means we turn on `AutoGenerateBindingRedirects` unless the
+developer has configured it explicitly.
+
+### Class Libraries
+
+Class libraries don't need binding redirects, unless they behave like apps, for
+instance unit testing projects.
+
+We generally don't want to compute and emit binding redirects because for most
+class library projects it's just wasting time. We need to figure out how we can
+detect unit testing projects during build and turn the setting on.
+
+## Q & A
+
+### Why is the setting not conditioned on the target framework?
+
+Over the last years we have not encountered a single instance where binding
```

Comment:
That is a bit bold. We just hit an issue with ValueTuple where having the binding redirect actually caused the application to fail when upgrading to 4.7.1 on the box. Granted it was a bug we fixed in the netstandard facade but non-the-less if there wasn't a binding redirect it would have worked just but with it it failed.

Reply:
But the problem isn't with Automatic Generation of Binding Redirects (i.e. false or incorrect generation of them). In other words, turning of ABRG wouldn't have fixed ValueTuple either.

Reply:
That's correct but my point is having binding redirects can cause some issues which is what I was point out. Mainly I think your statement may be a little strong and will likely get some feedback because of it.

Reply:
That's fair, all I'll try to reword. I think it's important to point out that ABRG isn't the culprit thought and that turning it on by default is safe